### DISCUSSION TITLE
OReplay data model

### DISCUSSION THREAD

**Discussion by jarufe** [oreplay/oreplay#6] [General]
Inspired by CPI system's data design

The link below allows users to edit the document. The goal is to share and to complete the information between all of the collaborators.
[Link to Google Diagrams](https://docs.google.com/drawings/d/1WoSdQBarJp15wpM_UacYFaErm5n-IbrCDXWzBJMZ5wM/edit?usp=sharing)

The link below is to share the complete breakout of the model. It should be, at this stage, a conceptual view of the data but I'm profiting of the final design of CPI model, so this is more a design than a model.

FIRST LINK, DON'T USE IT. It's the first draft of the model, kept to be able to compare the differences between approaches.
[Data Model, First Draft. Link to Google Docs](https://docs.google.com/document/d/1Hnsd8qT-zFoPzVLjk64N8vOPpiZBH9NtWfDTe0_j-wE/edit?usp=sharing)

Follow this discussion to find the reason for the change as this change is very important (fields that form the primary key because a dependence by existence are changed only to foreign keys in order to ease the use of the data)
[Data Model (Good). Link to Google Docs](https://docs.google.com/document/d/1p3olntx_UZ_V7gwnhz_sVVTuCzq6tCFIhFXJ7zG6YOM/edit?usp=sharing)

- Main difference: OReplay model covers Events having more than one stage. A CPI event is now an OReplay stage
- There is a strong dependence (by existence) between an event and its stages. Everything below this in the model is related to a stageq
- Stages can be one of several types: FootO, MTBO, Ski-O, Relay, Raid, Rogaine, Trail-O
- An event can be composed by stages of different types
- Radiocontrols or intermediate interesting points could be defined before the race. A control is attached to a class. The same control can be reused by different classes. The same control can be revisited by the same runner (individual races) or by the same team (runners from different legs in a relay)
- Controls can be of different types: clear, check, start, finish, control, readout
- Splits can be stored separately to have a breakdown of every punch, but it can also be the kind of breakdown used by XML IOF 3.0
- Runner can hold the complete data about a participant. Main identification data are: Name, BibNumber, license, SiCard, club, class, team (relays, raid and rogaine). It also can contain data about global times: check, clear, start, finish, race time, neutralization, etc. Also, data about stage results: time, points, bonus, penalty, etc.
- RunnerResults. For most of the common events, it seems to be a 1:1 relation with Runner, so this data could be stored in the Runner table. But it can be used to hold parts of the results in a stage. For example, this way we can cover Trail-O results. One occurrence could be the results of the normal controls, having the questions and answers stored in the ControlAnswer table and the Sportident data in the Split table. The rest of the occurrences could be the results of each timed group. Making this, Runner table could store the overall results of the event for a runner, using the first occurrence (the first stage) of the runner. So Runner could have fields to hold two kind of results: the overall results of the event, using the first occurrence (the first stage) of the runner; the results of the stage. Instead of having specific fields in Runner table, we could use one record of RunnerResults to store the overall of the event, one record to store the results of the stage, and other records for each partial result (like timed groups in Trail-O)
- Team and TeamResults could be similar as the previous one when talking about team competitions. The first occurrence (stage) of a team could store the overall results. Each occurrence of TeamResults could store the results of the current stage's sections
- It's very important to decide the scope of the solution that we want to provide. If the global results need to be computed from the raw data, then the model is not enough to store all the data needed. But, if the results are provided by the management software, then this can cover the needs
- Another important thing is to decide the deepness in the users' management. At first, we could think that we don't need much about users. If someone wants to manage an event, then he/she creates an event. But, we could extend the model in order to have users that manage several events through time. Even, we could think about an event administrator that creates another simple users to help in the management tasks. So, we could have a User table, an Event table, and a M:N relationship between them

**Comment by freefri** [comment]
We are talking about runners and teams, but it could be that our main entity is team. When we display the results, the only important entity is the whole team. I Adventure Racing 2+1, you do not care (in results) who punched what. You care about a team punching a CP as a team. For sure we need some extra useful information like Sportident card or gender, but this could be a new entity on top of the team entity (en team could be the main entity). In individual everts team name could be the competitor name, and the runner table could have one single entry with one Sportident card. In relay events, our team could be named Senior A and it could have 3 different runners assigned with 3 Sportident cards. Here, the results of each individual runner will be displayed depending on the splits of each Sportident card for each runner.

I think this cover all possible cases. In any case, the runner entity cold be extended if needed with the name and surname of each runner (and club or etc could also be relevant for raid).

It is true that similar solution will also work if the main entity is the runner, and we group runners in teams. This is maybe the approach used in sico (we should try to stick to similar data model as sico)

For the results platform, I think we should keep management to the minimum. If we do so, we will reduce the workload needed to operate the system by the Sportident timekeeper team.

I would suggest that the calculations within the results platform should be kept to the minimum, the timekeeping software OE, SiTiming or meos should calculate and validate the results. But oreplay should do some basic calculations. Do example when using radio controls. It would be great that quick download on Android in the middle of a race on rogaining or adventure racing uploads partial results to oreplay. The platform could do some basic scoring sum in order to generate partial results. Those results should be overwritten when the final download is performed after the finish line. In order to provide this functionality, CP entity should store a score field, and edit of this field should be managed before the event.

In footO radio controls should also update real time, in a similar way as in liveresults. In any case, with this type of partial results, the system should clearly display which results are real (calculated by oe) and which are partial (calculated by oreplay), and in case of partial results, oreplay should always asume the best scenario for the runner (e.g. even if intermediate CPs are missing, the whole linear race should be validated until the official download is done; if one runner of one team is missing in adverture racing, the whole team is validated; etc)

**Reply by freefri** [reply]
calculations should be done by timming software (not oreplay)

**Comment by jarufe** [comment]
Registration of data according to origin: independently, from SICO, from other federations
The objective was to imagine what the sequence or sequences through which we are going to feed ourselves with data would be like. A user comes here and does what to have an event, identify the competitors in their event, upload the data, etc. This way I could refine the data model to accommodate that way of working.

- Superuser users
  - Initially registered by the managers of OReplay
  - They will be able to register other users and classify them as superusers or administrators.

- Federations
  - In reality, it can be any type of organization that wants to have a continued and differentiated presence in OReplay.
  - Their registration is carried out by those responsible for OReplay at the request of the interested organization.
  - Starting there will be 2: FEDO (for SICO identifiers) and IOF (for OEventor identifiers)
  - An OReplay user may associate him with one or more federation identifiers. To do this he/she will have to know that identifier (searching OEventor, asking the FEDO, etc.)

- Users
  - Users can register themselves even if they do not initially manage events. They could do it, for example, to write their federation IDs and keep a history of their participations
  - Users who register themselves are neither administrators nor superusers.
  - A user will have the option to search and associate himself in events in which he has participated. After he find himself (finds the runner table record for the event) his userid will be written to the userid field of the runner table
    - An option of association can be automatic. If the user has indicated the UUID he has in a federation, all occurrences of runners in events in that federation and who have that UUID in the IdBD field could be searched. If he has entered UUIDs for multiple federations, the process could be repeated for all of them

- Management of an independent or minor event
  - If it is the first time, the user creates a user to manage the event
  - If you already have a user, log in
  - User creates an event
    - The person who creates an event is the event administrator.
    - He can associate other existing users (user ID must be provided) or create new users for other people to use and associate those users to the event
    - For simple management of minor single-day events of the classic type (Foot-O, MTBO, Ski-O), you can be given the option to indicate that it is going to be an event of that type and already create the event and one stage automatically
  - The user uploads start times. For example, with the desktop client, providing an XML file
    - Being independent, we cannot consider that data such as ID or License are covered. DBId may come, although we do not know if it will come from a persistent organized system (for example, FEDO's SICO). If we use XML, a person's data at most contains their name and an ID. Using OE CSV and knowing that certain data comes in the Num or Text fields, something else could be done, but it would depend on a selection by the user. Therefore, the DBId field can be filled out but it will not be assumed to be a SICO FEDO user code and consequently a possible OReplay user will not be searched for matching.
    - If the event has several stages and the data for all starting times is uploaded, the same runner ID could be created for the same person for all stages, since the complete identifier also contains event and stage
  - The user uploads results. For example, with the desktop client, providing an XML file of results (general, with a breakdown of partials or intermediate times with radiocontrols)
    - The runner will be searched and matched from the XML using one of the distinguishable fields available, probably the bib number

- Management of an event from SICO or other federations
  - The case is the same since these organizations are registered in the federation entity and their associates are identified with a specific UUID from their system.
  - The user, when creating the event, associates the federation from which the data will come from in the corresponding field of the event
  - When uploading start times, as the event has an associated federation value and the DBId field contains a value, when registering the runner you can search to see if there is an OReplay user with that federation identifier and that federation. If yes, the OReplay Id is assigned in the corresponding field of the runner table.

**Comment by jarufe** [comment]
**Videomeeting, 2024-01-08**
Some things to study / change:
- Raids. An event could be multi-stage but one class could be non-stop and the rest having several different stages. Check if the model covers this situation
- Team / Runner. As the information for the results is the same and the current XML generated by the software doesn`t cover a complete breakdown of the readings by the runners of a team, decission must be made about simplifying the model and keep only one entity for both situations (teams and individual competitions). 
  - The software stores the whole set of data. It's only that they don't export the whole set of data so far. I think that is better to keep the two entities.
  - Also, if we use radiocontrol data, we will be able to store data coming from some or all of the runners of a team.
- Naming convention. The other team members prefer to name entities and fields with a more comprehensive values, not codified as I proposed. No problem. It's necessary to make a proposal.
- Primary keys. Almost everything in the model depends on the existence of a stage (a class, a club, a course, a runner, etc), so the fields event id and stage id took part of the primary key of those tables at first (see first draft of the model at top of this discussion). Others prefer to make a more simple model, keep the tables with a 1-field primary key and make some operations related to the existence of the data by code. No problem (CPI model was that simple as well, but in this case I first preferred to follow the concept of the dependence by existence and to make a more complex primary keys)
- INTEGER type should be enough for autonumeric primary keys, but no problem to change them to BIGINT and have more range available.

**Comment by sergiogpajares** [comment]
## Proposal for changes in the data model

### Some missing points

**User's email** Users table doesn't contain an email that is required for logging or password recovery.

### Privacy and security

**Password storage** Passwords shouldn't be stored as a plain text in the data base for a basic reason. What happens if the DB get compromised for any reason? Thus, some encryption technique must be implemented. I am not a cybersecurity expert, in fact I know very little in this subject. However, after some research I found that the hashing algorithm `bcrypt` is recommended for this use case. Why not any other hashing algorithm? This one has been design targeting password storage. Moreover, it includes proper salt generation for enhanced security against rainbow attacks. Some other hashing algorithms could be considered.

`bcript` outputs a fixed length string of 60 characters no matter what the original password's length was. It contains all the information required to check the password (version number, number of encryption rounds, salt and hash). Thus, the password type should be `char(60)` instead of `varchar(128)`.

**alphanumeric identifiers for events** I believe that integer incremental identifiers should never be exposed to end-users. So, a possible attacker cannot loop over all the events etc... So, a fixed length random string should be used even if the app uses integer incremental ids. When uploading or querying for an event in the URL this string would be used. An example of this policy are `YouTUBE` videos.

**security token for data upload** A random string should be provided to make sure that only the right timekeepers can upload data to OReplay. An example of this policy is SportIdent Center `API key`.

**Two-factor authentication for super users** I think that a super user able to modify any user data or event setting would be a desired target for possible attackers. For that reason, two-factor authentication should be used. Maybe, we should not focus on this security measure during the first version, but in has to be consideren in the long run. I propose to use google authenticator. I don't know if we need to add anything to the user's table to implement this option but it should be studied.

This security policy could be considered mandatory for administrator users as well. But is would make things way more complicated for users and I would not make it mandatory at the beginning if we want a quick adoption of the APP. Any user should be given the option as an opt-in.

## UX design

**Public and hidden events** Event organizers that has not finished setting up their event should be able to mark it as hidden until desired. Thus, `is_public` is a boolean `default false` that should be added to the events table.

**Non-competitive events** OE supports non-competitive classes that display the results without positions and time differences. Also, it gives the possibility to display 'well done' instead of the time of the runner. Two bools `is-competitive` and `display_time-score` should be added to support this possibility.

**Filter events and/or clubs by country** A `country` and `region` enumerations should be added to those tables.

## Some statistical data

**Events scope** For analytical reasons, an enumerate with `event-scope` should be added to events table. The scope will be 'club', 'local', 'regional', 'national', 'international'. The same way winsplits storages this information.

**Users personal information** Many personal information data will be collected through the runner table. I wonder if this data should be stored in the user's table as well. User's will be able to edit this data. In the case that user's provided data and runners data (collected from timekeepers) differ, user's data will remain. This information can be useful to generate statistical reports of participation in an event to the event's administrator. This follows the line of the statistical report that FEDO ask their timekeepers to generate.

**Comment by freefri** [comment]
@jarufe I have made changes in my init sql file as discussed last week. The IDs are now `varchar(36)`. You can check it up in the commit 37346ba7249b95670094bb95e57b9b3a2e1dc5d5 

I have also made `stages.stage_type_id` NOT NULL as every other fk

**Reply by freefri** [reply]
also in this commit a new amend a7baf42f19b52f4afcbacc6b96cd174186876b4e

**Comment by freefri** [comment]
we need to change `time_seconds` from int to float (decimal or similar) and make sure the database timestamps works as well with miliseconds.

And also store timezone from the event in events table as discussed in https://github.com/oreplay/oreplay-backend/discussions/4#discussioncomment-8597521 @jarufe  @sergiogpajares

**Reply by jarufe** [reply]
I propose, to avoid loosing precission, to use `NUMERIC` data type. For example, `NUMERIC(9, 3)` would allow times for multiday events and keep the millisec fraction.
For the timezone there is already a field, not in the table of events but in stages. It's called `utc_value`. My idea was to store the offset of the utc zone (+01:00, -02:00, Z, etc), but another criteria could be followed. When creating an event, the frontend could propose the selection of an utc zone for the whole event and internally store the same value for all of the stages or give the chance to select a different zone for individual stages. Or, we could assume that the zone will be the same all of the days and move the field to the events table.

**Reply by freefri** [reply]
according to mysql documentation timestamp can have decimals https://dev.mysql.com/doc/refman/5.7/en/fractional-seconds.html 

I will change all timestamps to  `TIMESTAMP(3)` 

https://github.com/oreplay/oreplay-backend/commit/d337c4013aac20faecfedc444a824611cac37ff9 database timestamps and json ISO8601 should include miliseconds

**Comment by sergiogpajares** [comment]
It would also be advisable to have **event’s webpage** as a `nullable varchar` in `Events table` to be displayed to users in the frontend.

**Comment by freefri** [comment]
should we also remove all uuid colums from all tables? since we are using uuids as PK

**Reply by freefri** [reply]
[remove uuid fields from tables 63c062f](https://github.com/oreplay/oreplay-backend/commit/63c062f499978f61c432fba0b364f7857f6e1682)

**Reply by jarufe** [reply]
I've checked the model and found no fields to remove as they all have a meaning. 

users_federations: uuid field for an OReplay user that belongs to a federation means the identifier that the federation has provided to that person

clubs: uuid field is the way to identify the same club over all of the stages belonging to the same event. When importing data for the first stage, the club named XXX is assigned a PK-uuid. When importing data for the second stage, the same club is assigned another different PK-uuid. Having a non-PK-uuid lets the system link the same club over stages

courses: idem as clubs. Sometimes courses are completely different between stages, but often is the same description and the same association to classes so they can be compared, for example if a result is given summarizing courses instead of classes

classes: idem as courses. Normally, classes are the same between stages so the results can be summarized to get a global result. Having an extra non-PK-uuid field lets the system link the same class over stages

teams: uuid field has the same meaning as the previous ones. class_uuid is a denormalized value. If we have not only the FK of the class but also de extra value that links the same class over stages, probably it will be useful to compose the viewing of the results

runners: idem as teams

runner_results: idem as teams and runners for class_uuid and also for runner_uuid

team_results: idem as runner_results but for class_uuid and team_uuid

**Reply by freefri** [reply]
@jarufe What you say makes sense, let's discuss this on our next meeting, because we need an strategy to make this matching (when and how)

**Comment by jarufe** [comment]
## Stage types, disciplines, modalities. Proposal for change in data model

### stage_types not enough. Splitting the concept

At this moment, only one concept (stage_types) tried to cover different kind of treatments when the disciplines or the way of managing results were diverse. It doesn´t separate if the stage is Foot-O or MTBO and this is needed when providing more information (icons, for example) in the front-end

**IOF and FEDO statutes** regulate the disciplines covered by those federations. So, one concept could be that of discipline and then keep another (if useful) for special data treatment procedures

### stage_disciplines

Having in mind the disciplines stated in FEDO's statutes: Foot-O, MTBO, Ski-O, Trail-O, Raid and Ultrascore-Rogaine

### stage_displays

Valid values could be: Interval Start, Sprint, Score, Mass Start, Chase Start, Relay, Pre-O, Temp-O

**Reply by sergiogpajares** [reply]
In the `stage_type` as a TrailO competition it shall be added `TrailO_Sprint` or any other proposed naming. I also propose change name `classic` for `Interval Start`. Also, `Score` shall be added.

Name `stage_type` can be substituted by `stage_display`.

In the front-end, some logic can be added to display only valid stage_types. If you choose TrailO, only Pre-O, Temp-O and Sprint. If you choose Foot-O, only Interval, Mass or Chase Start should be showed along with Score.

**Reply by sergiogpajares** [reply]
Considering that time bonus is not a common feauture nor penalty in regular foot-o events. The API should not give those in the foot-o events.

**Comment by jarufe** [comment]
## Splits table model changed. Added two fields

**time_seconds** NUMERIC(9, 3)
**position** INTEGER

The exported files include the time in seconds (could have decimals if fractions of seconds are used) from the beginning of the runner's race. Some of the exported files also include the position of the runner in that split control

Previous fields are maintained. They came from the CPI (radiocontrol) model where the real date/time value and it conversion to milliseconds were stored (reading_time TIMESTAMP; reading_milli BIGINT)

**Comment by jarufe** [comment]
## Runners and Teams tables model changed. Added one field

**course_id** VARCHAR(36)

Now dealing with relay format, each runner usually has an specific course. Added the same field to the team table in case another kind of competition assigns a different course to each team.
